include/cetl/span.h

```diff
@@ -1 +1,772 @@
+    }
+
+    /// Iterator to the address after the last element in the span.
+    /// @code{.unparsed}
+    ///                     end
+    ///                      v
+    ///    +-----------------------+
+    ///    | 0 | 1 | 2 | 3 | x | x |
+    ///    +-----------------------+
+    ///
+    /// @endcode
+    /// @return normal iterator.
+    /// @see std::span::end
+    constexpr iterator end() const noexcept
+    {
+        return iterator(&data_[Extent]);
+    }
+
+    /// Reverse iterator to the last element in the span.
+    /// This is the same as span::rend if the span size is 0.
+    /// @return reverse iterator after the end of the span.
+    /// @see std::span::rbegin
+    constexpr reverse_iterator rbegin() const noexcept
+    {
+        return reverse_iterator(end());
+    }
+
+    /// Reverse iterator to the address before the first element in the span.
+    /// @code{.unparsed}
+    /// rend
+    ///  v
+    ///    +-----------------------+
+    ///    | 0 | 1 | 2 | 3 | x | x |
+    ///    +-----------------------+
+    ///
+    /// @endcode
+    /// @return reverse iterator from the begining of the span.
+    /// @see std::span::rend
+    constexpr reverse_iterator rend() const noexcept
+    {
+        return reverse_iterator(begin());
+    }
+
+    ///@}
+    // +----------------------------------------------------------------------+
+    ///@{ @name Element Access
+    // +----------------------------------------------------------------------+
+
+    ///
+    /// Returns a reference to the first element.
+    /// Calling this method on an empty span is undefined.
+    /// @return Reference to the first element. This is the same as `*obj.begin()`.
+    /// @see std::span::front
+    constexpr reference front() const
+    {
+        CETL_DEBUG_ASSERT((Extent != 0), "Calling front on an empty span is undefined.");
+        return *begin();
+    }
+
+    ///
+    /// Returns a reference to the last element.
+    /// Calling this method on an empty span is undefined.
+    /// @return Reference to the last element. This is the same as `*(obj.end() - 1)`.
+    /// @see std::span::back
+    constexpr reference back() const
+    {
+        CETL_DEBUG_ASSERT((Extent != 0), "Calling back on an empty span is undefined.");
+        return *(end() - 1);
+    }
+
+    ///
+    /// Provides access to the internal data the span is a view into.
+    /// @return A pointer to the begining of the span sequence.
+    /// @see std::span::data
+    constexpr pointer data() const noexcept
+    {
+        return data_;
+    }
+
+    ///@}
+    // +----------------------------------------------------------------------+
+    ///@{ @name Observers
+    // +----------------------------------------------------------------------+
+
+    ///
+    /// If the span has a zero size or not.
+    /// @return true if the span size is 0 where "size" is the same as span::extent
+    /// for this specialization.
+    [[nodiscard]] constexpr bool empty() const noexcept
```

Comment:
Was wondering when I'd see the first `[[nodiscard]]`

Reply:
Yeah, this is a smart move on LLVM's part (not sure if gcc followed suit). A lot of people think this is a verb. This makes it more probable that the compiler warning about not consuming the result will clue the user in.

Reply:
nodiscard is not supported in C++14 so should not be used here unless made conditional on the standard version

Reply:
ah crud. You're right. I think it's best to omit this since it doesn't affect the ABI in any way. Or should we add a CETL_NODISCARD macro? But macros are bad.

Reply:
I'm proposing a slightly different approach where we do use macros but we explicitly put the feature testing in the header. Continuing this pattern will prevent cetl.h from becoming a uber portability layer and it lets the user of these types see, very clearly, what adaptations our types make based on the standard in use. Let me know what you think.

Reply:
I am not sure I understand your proposal in its entirety but it sure sounds great.

Reply:
see my latest push

Reply:
We good with the latest @pavel-kirienko ?

Reply:
It works, but I suspect it might get cumbersome further down the road when we add more optinoal capabilities like `[[noinline]]` etc. I am fine with either approach but all other things equal I would incline towards `CETL_NODISCARD`.